x/msg_authorization/internal/keeper/keeper.go

```diff
@@ -0,0 +1,41 @@
+package keeper
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/msg_authorization/internal/types"
+	"time"
+)
+
+type Keeper struct {
+	storeKey sdk.StoreKey
+	cdc      *codec.Codec
+	router   sdk.Router
+}
+
+// NewKeeper constructs a message authorisation Keeper
+func NewKeeper(storeKey sdk.StoreKey, cdc *codec.Codec, router sdk.Router) Keeper {
+	return Keeper{
+		storeKey: storeKey,
+		cdc:      cdc,
+		router:   router,
+	}
+}
+
+func (k Keeper) DispatchActions(ctx sdk.Context, grantee sdk.AccAddress, msgs []sdk.Msg) sdk.Result {
+	//TODO
+	return sdk.Result{}
+}
+
+func (k Keeper) Grant(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress, capability types.Capability, expiration time.Time) {
```

Comment:
I would love to see an implementation.

One question is how you handle if I grant a Capability when there is already another capability of the same type registered? (This is for you @aaronc). Revoke will remove by type, so I assume we only store one per type. We can either overwrite, or fail (forcing explicit Revoke then Grant). The first requires fewer Tx for the user, the later is more explicit and avoids unintended actions.

Reply:
My opinion is it should fail, and we let the user know. I don't see why it would be intended to grant a new capability of same msg type

Reply:
My intention and initial implementation was actually to overwrite - the logic being that a granter may want to frequently update their grants for instance extending expiration times, increasing spend allowances, etc. So with that in mind I thought it simplest just overwrite.

Reply:
Do either of you see any major issue with allowing that behavior?

Reply:
It should be very well documented, both in the adr, spec and the cli tool